From 117fb8844308f86732d3910b1797528bf9a9744c Mon Sep 17 00:00:00 2001
From: Karim Pinter <karim.pinter@qt.io>
Date: Thu, 4 Sep 2025 14:52:36 +0300
Subject: [PATCH 106/553] Work around QFileSystemEngine behavior on VxWorks
 hostFs

On VxWorks with hostFs (commonly used for debugging via Workbench),
using QFile::exists() triggers a statfs() call, which fails. This change
uses stat() check instead, which behaves more reliably in this context.
Additionally, VxWorks stat() still incorrectly reports that the file has
been deleted. This fix accounts for that false negative and avoids
treating the file as missing.

Task-number: QTBUG-139007
Pick-to: 6.9 6.8
Change-Id: Ie2cd277868e0d57bb8288271c8973da17162eef3
Reviewed-by: Thiago Macieira <thiago.macieira@intel.com>
(cherry picked from commit b4c59c7d49e8322c552e87751b12a7deb3ce349b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/io/qfilesystemengine_unix.cpp | 44 +++++++++++++----------
 1 file changed, 26 insertions(+), 18 deletions(-)

diff --git a/src/corelib/io/qfilesystemengine_unix.cpp b/src/corelib/io/qfilesystemengine_unix.cpp
index 32818949016..477a0892d92 100644
--- a/src/corelib/io/qfilesystemengine_unix.cpp
+++ b/src/corelib/io/qfilesystemengine_unix.cpp
@@ -1026,28 +1026,36 @@ bool QFileSystemEngine::fillMetaData(const QFileSystemEntry &entry, QFileSystemM
     // third, we try access(2)
     if (what & (QFileSystemMetaData::UserPermissions | QFileSystemMetaData::ExistsAttribute)) {
 #if defined(Q_OS_VXWORKS)
-        // on VxWorks if the filesystem is not POSIX, access() always returns false, despite the
-        // file is readable
         struct statfs statBuf;
-        if (statfs(nativeFilePath, &statBuf) != 0) {
-            what &= ~QFileSystemMetaData::LinkType; // don't clear link: could be broken symlink
-            data.clearFlags(what);
-            return false;
-        }
-        if (statBuf.f_type != NFSV2_MAGIC && statBuf.f_type != NFSV3_MAGIC &&
-            statBuf.f_type != HRFS_MAGIC) {
+        if (statfs(nativeFilePath, &statBuf) == 0) {
+            if (statBuf.f_type != NFSV2_MAGIC && statBuf.f_type != NFSV3_MAGIC &&
+                statBuf.f_type != HRFS_MAGIC) {
 #if __has_include(<dosFsLib.h>)
-            if (data.entryFlags & QFileSystemMetaData::OwnerWritePermission) {
-                data.entryFlags |= QFileSystemMetaData::UserWritePermission;
-            }
-            if (data.entryFlags & QFileSystemMetaData::OwnerExecutePermission) {
-                data.entryFlags |= QFileSystemMetaData::UserExecutePermission;
-            }
+                if (data.entryFlags & QFileSystemMetaData::OwnerWritePermission) {
+                    data.entryFlags |= QFileSystemMetaData::UserWritePermission;
+                }
+                if (data.entryFlags & QFileSystemMetaData::OwnerExecutePermission) {
+                    data.entryFlags |= QFileSystemMetaData::UserExecutePermission;
+                }
 #endif
-            data.entryFlags |= QFileSystemMetaData::UserReadPermission |
-                    QFileSystemMetaData::ExistsAttribute;
-            return true;
+                data.entryFlags |= QFileSystemMetaData::UserReadPermission |
+                        QFileSystemMetaData::ExistsAttribute;
+                return true;
+            }
         }
+#if defined(QT_DEBUG)
+        else {
+              //on VxWorks hostfs, used for debugging, failes on statfs and falsely reports
+              //WasDeleted
+              statResult = QT_STAT(nativeFilePath, &statBuffer);
+              if (statResult == 0) {
+                  data.entryFlags |= QFileSystemMetaData::UserReadPermission |
+                                     QFileSystemMetaData::ExistsAttribute;
+                  data.entryFlags &= ~QFileSystemMetaData::WasDeletedAttribute;
+                  return true;
+              }
+        }
+#endif
 #endif
         // calculate user permissions
         auto checkAccess = [&](QFileSystemMetaData::MetaDataFlag flag, int mode) {
-- 
2.51.2

